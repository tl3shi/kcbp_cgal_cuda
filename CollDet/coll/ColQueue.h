
//---------------------------------------------------------------------------
//                              Queue
//---------------------------------------------------------------------------
//  Copyright (C):
//---------------------------------------------------------------------------
//CVSId: "@(#)$Id: ColQueue.h,v 1.5 2004/04/29 13:54:20 weller Exp $"


#ifndef Queue_H
#define Queue_H
#if defined(__sgi) || defined(_WIN32)
#pragma once
#endif

//---------------------------------------------------------------------------
//  Includes
//---------------------------------------------------------------------------

#include <vector>

#include <OpenSG/OSGConfig.h>
#include <OpenSG/OSGLock.h>

#include <ColExceptions.h>


namespace col {

//---------------------------------------------------------------------------
//  Forward References
//---------------------------------------------------------------------------

//---------------------------------------------------------------------------
//   Types
//---------------------------------------------------------------------------


//---------------------------------------------------------------------------
//  The Class
//---------------------------------------------------------------------------

template <typename T>
class Queue
{


/***************************************************************************/
public:

//---------------------------------------------------------------------------
//  Public Instance methods
//---------------------------------------------------------------------------

/*------------------ constructors & destructors ---------------------------*/

	Queue();
	virtual ~Queue() throw();

/*-------------------------- access --------------------------------*/

	void	add( const T &element );
	T*		remove();
	void	swap();
	int		back_size() const;
	int		front_size() const;

//---------------------------------------------------------------------------
//  Public Class methods
//---------------------------------------------------------------------------




/***************************************************************************/
protected:

//---------------------------------------------------------------------------
//  Protected Instance methods
//---------------------------------------------------------------------------

	void	init();

//---------------------------------------------------------------------------
//  Instance variables
//---------------------------------------------------------------------------

	OSG::Lock		 *m_write_lock;		// for preventing simul. writing
	std::vector<T> 	m_elem[2];			// the elements
	std::vector<T>	 *m_front, *m_back;	// pointers to elem[i]
	unsigned int	 m_back_top;			// next valid element in back queue
	unsigned int	 m_front_size, m_back_size;	// # valid elements

//---------------------------------------------------------------------------
//  Class variables
//---------------------------------------------------------------------------

	static int M_Num_queue;		// queue counter


/***************************************************************************/
private:

//---------------------------------------------------------------------------
//  Private Instance methods
//---------------------------------------------------------------------------

	// prohibit copy constructor (move to 'public' if you need one)
	Queue( const Queue &source );

	// prohibit assignment
	Queue& operator = ( const Queue &source );


//---------------------------------------------------------------------------
//  Private Class methods
//---------------------------------------------------------------------------


};


} // namespace col

#include <ColQueue.hpp>


#endif /* Queue_H */

