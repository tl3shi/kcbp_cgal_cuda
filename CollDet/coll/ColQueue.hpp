
/*****************************************************************************\
 *                              Queue
\*****************************************************************************/
/*! @file 
 *
 *  @brief
 *    Class Queue and exceptions it might throw.
 *
 *  @author Gabriel Zachmann
 *  
 */



//---------------------------------------------------------------------------
//  Includes
//---------------------------------------------------------------------------


#include <stdlib.h>
#include <stdio.h>

#include <OpenSG/OSGThreadManager.h>

namespace col {


/***************************************************************************\
 *                            Public Types                                 *
\***************************************************************************/




/***************************************************************************\
 *                      Private/Protected Instance variables               *
\***************************************************************************/


#define MaxNumQueues 1000


/** @class Queue
 *
 * A double-buffered FIFO queue, so that concurrent threads can write and read.
 *
 * The queue is implemented for efficiency: memory is never released
 * (except, of course, when a queue is destructed).
 *
 * Multiple concurrent threads (producers) can write into the front buffer of a
 * Queue, while another concurrent thread (consumer) can read from the back
 * buffer.  Thus, producers and consumer do not block each other.
 *
 * Although this class is in the "col" namespace, it is quite general and
 * should be useful in other areas, too.
 *
 * @warning
 *   Concurrent writing and simultaneous writing and swapping
 *   are multithread-safe.
 *   All other access to the back-end (like reading) is @e not multithread-safe.
 *   There should be only one consumer at the back-end.
 *   If you need more thread-safety, just derive a sub-class.
 *
 * @warning
 *   Only the consumer should swap the buffers!
 *   (Reading and swapping are @e not thread-safe.)
 *
 * @warning
 *   Instances can be created only after osg::Init()!
 *
 * @see
 *
 * @implementation
 *   I use an STL vector instead of an STL queue, because I hope that
 *   vector doesn't shrink, in order to increase performance.
 *   In addition, I make sure that vector->size() never decreases.
 *
 * @implementation
 *   By restricting the number of consumers to one only,
 *   the number of lockings can be reduced to a minimum.
 *
 * @implementation
 *   Queue depends on OSG only for the lock support.
 *   We could get independent from OSG by writing portable locks ourselves.
 *
 **/

/***************************************************************************\
 *                         Public Instance Methods                         *
\***************************************************************************/


/** @name               Constructors and desctructors
 */
//
// @{



/**  Construct queue
 *
 **/

template <typename T>
Queue<T>::Queue()
{
	init();
}


template <typename T>
Queue<T>::~Queue() throw()
{
    // decrease ref count fore _write_lock
    osg::subRefP(m_write_lock);
    //	OSG::ThreadManager::the()->removeLock( _write_lock );
}


/**  Initializes a queue
 *
 * @throw XNoLock
 *   If OSGThreadManager couldn't create a lock.
 * @throw XTooMany
 *   If too many queues were requested.
 *
 * Meant to be called by constructors.
 *
 **/

template <typename T>
void Queue<T>::init()
{
	if ( M_Num_queue >= MaxNumQueues )
		throw XQueueTooMany( "queues" );

	char name[100];
	sprintf( name + strlen("Queue_"), "%4d", M_Num_queue );
	m_write_lock = OSG::ThreadManager::the()->getLock(name);
	if ( ! m_write_lock )
		throw XQueueNoLock();
    // increase ref count for _write_lock
    osg::addRefP(m_write_lock);

	m_front = &m_elem[0];
	m_back  = &m_elem[1];
	m_back_top = 0;
	m_front_size = m_back_size = 0;

	M_Num_queue ++ ;
}




// @}
//
/** @name                            Access
 */
//
// @{


/**  Copy @p element to the end of the front queue
 *
 * @param element	Element to be copied into queue
 *
 * Access is exclusive, so that multiple threads can add elements.
 * Elements are copied, so the caller can destroy them after add().
 *
 * @throw XTooMany
 *   If more than maxsize elements are added to the front before
 *   a swap() occurs. Maxsize is set through the constructor.
 *
 **/

template <typename T>
void Queue<T>::add( const T &element )
{
	m_write_lock->aquire();
	try
	{
		if ( m_front_size < m_front->size() )
			// overwrite
			(*m_front)[m_front_size++] = element;
		else
		{
			m_front->push_back( element );
			m_front_size = m_front->size();
		}
	}
	catch ( ... )
	{
		m_write_lock->release();
		throw;
	}
	m_write_lock->release();
}



/**  Return pointer to current front element of queue
 *
 * @return
 *   Pointer to front element, or NULL, if the queue has been exhausted.
 *
 * This function also removes the front element conceptually.
 * The pointer remains valid until the next @c swap.
 *
 * @warning
 *   This function is @e not thread-safe!
 *   Only one thread should read from a Queue.
 *
 * @implementation
 *   For efficiency, elements are not really destroyed.
 *
 **/

template <typename T>
T* Queue<T>::remove()
{
	if ( m_back_top >= m_back->size() )
		return NULL;
	return & (*m_back)[m_back_top++];
}



/**  Number of elements in the back queue
 *
 * @return
 *   Return number of elements remaining in the back queue.
 *
 * @warning
 *   This number decreases as you remove() elements from the back queue!
 *
 **/

template <typename T>
int Queue<T>::back_size() const
{
	return m_back_size - m_back_top;
}


/**  Number of elements in the front queue
 *
 * @return
 *   Return number of elements in the front queue
 *
 * This number increases as you add() elements to the front queue.
 *
 **/


template <typename T>
int Queue<T>::front_size() const
{
	return m_front_size;
}


/**  Swap front and back queue
 *
 * The front and the back queue are exchanged.
 * The new front (formerly back) is cleared.
 *
 * @warning
 *   This function is should be called only by the consumer;
 *   and there should be only @e one consumer.
 *
 **/


template <typename T>
void Queue<T>::swap()
{
	std::vector<T> *vtmp;

	m_write_lock->aquire();

	vtmp = m_front;  m_front = m_back;  m_back = vtmp;
	m_back_size = m_front_size;
	m_back_top = 0;
	m_front_size = 0;

	m_write_lock->release();
}


// @}



/***************************************************************************\
 *                        Class variables                                  *
\***************************************************************************/


// number of queues already created
template <typename T>
int Queue<T>::M_Num_queue = 0;


/***************************************************************************\
 *                      Public Class methods                               *
\***************************************************************************/





/**  Zum Abpasten fuer eine Funktion, Methode, oder Makro-"Funktion".
 *
 * @param param1	Comment for param1
 *
 * @return
 *   Text for return value.
 *
 * Detaillierte Beschreibung ...
 *
 * @throw Exception
 *  Beschreibung ..
 *
 * @warning
 *   Dinge, die der Aufrufer unbedingt beachten muss...
 *
 * @assumptions
 *   Annahmnen, die die Funktion macht...
 *
 * @sideeffects
 *   Nebenwirkungen, globale Variablen, die veraendert werden, ..
 *
 * @todo
 *   Was noch getan werden muss
 *
 * @bug
 *   Bekannte Bugs dieser Funktion
 *
 * @see
 *   ...
 *
 * @implementation
 *   Implementierungsdetails, TODOs, ...
 *
 **/




} // Col


