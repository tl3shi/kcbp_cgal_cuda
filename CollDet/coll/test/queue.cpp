
/*****************************************************************************\
 *                          Testprogram for Queue
\*****************************************************************************/

/*! @file
 *
 *  @brief
 *    Testprogram for Queue
 *
 *  Tests:
 *    - Single threaded: preservation of number of elements, order of elements,
 *      elements are correct, 2x swap clears queues.
 *    - 2-threaded, 1 producer, 1 consumer: correct number of elements,
 *      enough elements afterthe first swap, sanity check of elements.
 *    - 3 producers: number of elements.
 *    - 15 producers: dito.
 *
 *  @implementation:
 *    I use printf instead of cout, because cout is not thread-safe!
 *
 *  @author Gabriel Zachmann
 *
 *  @todo
 *    Die Verwendung und Deklaration von @c printerr() scheint inkonsistent!
 */



//---------------------------------------------------------------------------
//  Includes
//---------------------------------------------------------------------------


#include <stdlib.h>
#include <stdio.h>

#include <ColQueue.h>

#include <OpenSG/OSGBaseFunctions.h>
#include <OpenSG/OSGThread.h>
#include <OpenSG/OSGThreadManager.h>
#include <OpenSG/OSGBarrier.h>
#include <OpenSG/OSGLog.h>


//---------------------------------------------------------------------------
//  Types
//---------------------------------------------------------------------------

struct ThreadData
{
	col::Queue<int> *qu;
	int millisec;
	int nthreads;
	int n_added;
	int nloops;
};

//---------------------------------------------------------------------------
//  Global variables
//---------------------------------------------------------------------------

const int nelems = 100;
const char barrier_name[] = "queue";

int elem[nelems];
int num_errs = 0;
int exception_occured = 0;


/***************************************************************************\
 *                                 Misc                                    *
\***************************************************************************/


void printerr( int *nerrs = NULL )
{
	static int n = 0;
	static int test_n = 0;

	if ( nerrs )
	{
		// another test section finished
		printf("test %d finished (with %d errors).\n", test_n, n );
		(*nerrs) += n;
		n = 0;
		test_n ++ ;
	}
	else
	{
		printf("*** %d. test failed!\n", test_n );
		n ++ ;
	}
}



void threadfunc( void *thread_parm )
{
	ThreadData *td = static_cast<ThreadData*>(thread_parm);
	
	try
	{
		//printf("threadfunc started");

		// fill queue
		for ( int j = 0; j < td->nloops; j ++ )
		{
			osg::osgsleep(td->millisec);
			for ( int i = 0; i < nelems; i ++ )
				td->qu->add( elem[i] );				// could throw
			td->n_added += nelems ;
		}

		if ( td->qu->front_size() == td->nloops*nelems && td->millisec > 10 )
		{
			printf("queue: threadfunc: hm .. still all elements in the front "
				   "queue after %d milliseconds!", td->nloops*td->millisec );
		}
	}
	catch ( col::XQueueTooMany )
	{
		exception_occured = 1;
	}

	// rendez-vous with main
	osg::Barrier *barrier = osg::ThreadManager::the()->getBarrier(barrier_name);
	if ( ! barrier )
		fprintf(stderr,"queue: threadfunc: getBarrier failed!\n");
	else
		barrier->enter(td->nthreads);
}

/***************************************************************************\
 *                                 Main                                    *
\***************************************************************************/


int main( int argc, char *argv[] )
{
	int *e;


	osg::osgLog().setLogLevel(osg::LOG_WARNING);
	osg::osgInit(argc, argv);

	col::Queue<int> qu;

	for ( int i = 0; i < nelems; i ++ )
		elem[i] = rand();			// should produce same sequence every time

	// single thread

	for ( int i = 0; i < nelems; i ++ )
		qu.add( elem[i] );

	// back queue should be empty
	e = qu.remove();
	if ( e )
	{
		printerr(false);
		puts("back queue doesn't seem to be empty!");
	}
	printerr(&num_errs);				// 0. test finished

	// swap
	qu.swap();

	// check size
	if ( qu.back_size() != nelems )
	{
		printerr(false);
		printf("back_size (%d) != nelems (%d)!\n", qu.back_size(), nelems );
	}
	printerr(&num_errs);				// 1. test finished

	qu.swap();

	// back queue should be empty
	e = qu.remove();
	if ( e )
	{
		printerr(false);
		puts("back queue doesn't seem to be empty!");
	}
	printerr(&num_errs);				// 2. test finished
	if ( qu.back_size() != 0 )
	{
		printerr(false);
		printf("back_size (%d) != 0!\n", qu.back_size() );
	}
	printerr(&num_errs);				// 3. test finished

	for ( int i = 0; i < nelems; i ++ )
		qu.add( elem[i] );

	qu.swap();

	// get all elems from back queue
	for ( int i = 0; i < nelems; i ++ )
	{
		// add to front while removing from back
		qu.add( elem[nelems-1-i] );

		e = qu.remove();
		if ( ! e )
		{
			printerr(false);
			printf("received NULL at position %d!\n", i );
			break;
		}

		if ( *e != elem[i] )
		{
			printerr(false);
			printf("position %d is different after queue!", i );
			printf("input to queue:");
			for ( int j = 0; j < nelems; j ++ )
				printf("%4d ", elem[i] );
			putchar('\n');
			printf("output from queue starting with position %d:", i );
			printf("%4d ", *e );
			for ( int j = i+1; j < nelems; j ++ )
				printf("%4d ", *qu.remove() );
			putchar('\n');

			break;
		}
	}
	printerr(&num_errs);				// 4. test finished

	// back is empty
	if ( qu.back_size() != 0 )
	{
		printerr(false);
		puts("back queue size doesn't seem to be empty!");
	}
	printerr(&num_errs);				// 5. test finished

	// front is full again
	qu.swap();
	if ( qu.back_size() != nelems )
	{
		printerr(false);
		printf("back_size (%d) != nelems (%d)!\n", qu.back_size(), nelems );
	}
	printerr(&num_errs);				// 6. test finished

	for ( int i = 0; i < nelems; i ++ )
	{
		e = qu.remove();
		if ( ! e )
		{
			printerr(false);
			printf("received NULL at position %d!\n", i );
			break;
		}

		if ( *e != elem[nelems-1-i] )
		{
			printerr(false);
			printf("position %d is different after queue!\n", nelems-1-i );
			break;
		}
	}
	printerr(&num_errs);				// 7. test finished

	// 2 threads, 1 producer, 1 consumer

	osg::BaseThread *thread = osg::ThreadManager::the()->getThread(NULL);

	if ( ! thread )
	{
		fprintf(stderr,"queue: ThreadManager didn't return a thread!\n");
		return 1;
	}

	osg::Barrier *barrier = osg::ThreadManager::the()->getBarrier(barrier_name);
	if ( ! barrier )
	{
		fprintf(stderr,"queue: main thread: getBarrier failed!\n");
		return 1;
	}

	ThreadData td;
	td.qu = &qu;
	td.millisec = 300;
	td.nthreads = 2;
	td.n_added = 0;
	td.nloops = 5;

	bool ok;
	ok = thread->runFunction( threadfunc, &td );

	if ( ! ok )
	{
		fprintf(stderr,"queue: running thread failed!\n");
		return 1;
	}
	// wait some time to start the thread
	osg::osgsleep(400);

	// we are the back end
	
	const int sl = 480;
	int n_removed = 0;

	for ( int i = 0; i < td.nloops; i ++ )
	{
		osg::osgsleep(sl);
		if ( i == td.nloops-1 )
		{
			// last loop; rendez-vous with threadfunc,
			// so we get all elements still to be transmitted from producer
			barrier->enter(td.nthreads);
		}

		int n = qu.front_size();
		if ( n == 0 && i == 0 )
		{
			printf("queue: hm .. no elements in the front queue after %d "
				   " milliseconds!\n", sl );
			puts(" this should not happen!");
			continue;
		}

		qu.swap();

		//printf("back_size = %d\n", qu.back_size() );
		// don't print this for automated regression test, because
		// the actual num. of elements in the queue might vary due to different
		// timings on different platforms (race condition)

		if ( qu.back_size() < n )
		{
			printerr(false);
			printf("back queue size (%d) < front queue size (%d) after swap!\n",
				   qu.back_size(), n );
			break;
		}

		n_removed += qu.back_size();

		for ( int j = 0; j < qu.back_size(); j ++ )
		{
			e = qu.remove();
			if ( ! e )
			{
				printerr(false);
				printf("received NULL at position %d!\n", i );
				break;
			}

			if ( *e < 0 )
			{
				printerr(false);
				printf("got element (%d) < 0!\n", *e );
				break;
			}

			// is there anything else I could check?
			// any post-condition which is true even with multiple threads?
		}
	}

	if ( td.n_added != n_removed )
	{
		printerr(false);
		printf("#elements retrieved from back (%d) != "
			   "#elements added by thread (%d)!\n", n_removed, td.n_added );
	}

	printerr(&num_errs);				// 8. test finished

	// many threads, n producer, 1 consumer

	const int nthreads = 17;
	osg::BaseThread *threads[nthreads];
	for ( int i = 0; i < nthreads; i ++ )
	{
		threads[i] = osg::ThreadManager::the()->getThread(NULL);
		if ( ! threads[i] )
		{
			fprintf(stderr,"queue: ThreadManager didn't return %d-th thread!\n",
					i );
			return 1;
		}
	}

	// first, a simple case
	td.millisec = 0;
	td.nthreads = 4;
	td.n_added = 0;
	for ( int i = 0; i < td.nthreads-1; i ++ )
	{
		ok = threads[i]->runFunction( threadfunc, &td );
		if ( ! ok )
		{
			fprintf(stderr,"queue: running thread %d failed!\n", i );
			return 1;
		}
	}

	barrier->enter(td.nthreads);	// wait for other threads to finish

	if ( td.n_added != qu.front_size() )
	{
		printerr(false);
		printf("queue: %d threads have lost something (%d vs. %d elements",
				td.nthreads-1, td.n_added, qu.front_size() );
	}

	qu.swap();
	printerr(&num_errs);				// 9. test finished

	// now, throw a lot of snowballs at the queue
	td.millisec = 0;
	td.nthreads = 9;
	td.n_added = 0;
	for ( int i = 0; i < td.nthreads-1; i ++ )
	{
		ok = threads[i]->runFunction( threadfunc, &td );
		if ( ! ok )
		{
			fprintf(stderr,"queue: running thread %d failed!\n", i );
			return 1;
		}
	}

	barrier->enter(td.nthreads);	// wait for other threads to finish

	if ( td.n_added != qu.front_size() )
	{
		printerr(false);
		printf("queue: %d threads have lost something (%d vs. %d elements\n",
				td.nthreads-1, td.n_added, qu.front_size() );
	}

	qu.swap();

	if ( td.n_added != qu.back_size() )
	{
		printerr(false);
		printf("queue: swap has lost something "
				"(after %d threads filled front; %d vs. %d elements\n",
				td.nthreads-1, td.n_added, qu.back_size() );
	}
	else
		printf("%d threads have added %d elements.\n",
			   td.nthreads-1, td.n_added );

	qu.swap();
	printerr(&num_errs);				// 10. test finished

	if ( num_errs )
		printf("queue: %d tests failed!\n", num_errs );
	else
		puts("queue: all tests ok.");
	
	return 0;
}


