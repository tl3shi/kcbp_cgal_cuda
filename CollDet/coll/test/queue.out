test 0 finished (with 0 errors).
test 1 finished (with 0 errors).
test 2 finished (with 0 errors).
test 3 finished (with 0 errors).
test 4 finished (with 0 errors).
test 5 finished (with 0 errors).
test 6 finished (with 0 errors).
test 7 finished (with 0 errors).
test 8 finished (with 0 errors).
test 9 finished (with 0 errors).
8 threads have added 4000 elements.
test 10 finished (with 0 errors).
queue: all tests ok.
